### PR TITLE
changes condition status to be a string, rather than a boolean

### LINKED ISSUE #13074: Operator CRD status incompatible with kstatus
### Describe the bug

When using the new operator with Flux, the create Keycloak CRD instance never reaches ready status because the reported status deviates from the convention otherwise used: It uses yaml booleans instead of yaml strings.

(also see [kstatus](https://github.com/kubernetes-sigs/cli-utils/tree/master/pkg/kstatus)).

The conventions seems to be yaml strings:

Certificate from cert-manager:
```yaml
status:
  conditions:
    - lastTransitionTime: '2022-07-13T06:51:51Z'
      message: Certificate is up to date and has not expired
      observedGeneration: 1
      reason: Ready
      status: 'True'
      type: Ready
  notAfter: '2022-10-11T06:51:51Z'
  notBefore: '2022-07-13T06:51:51Z'
  renewalTime: '2022-09-11T06:51:51Z'
  revision: 1
```

SealedSecret from bitnami:
```
status:
  conditions:
    - lastTransitionTime: '2022-07-13T06:51:20Z'
      lastUpdateTime: '2022-07-13T06:51:20Z'
      status: 'True'
      type: Synced
```

### Version

18.0.2

### Expected behavior

The Keycloak CRD should report its status using yaml strings:

```yaml
status:
  conditions:
    - message: ''
      status: 'True'
      type: Ready
    - message: ''
      status: 'False'
      type: HasErrors
    - message: ''
      status: 'False'
      type: RollingUpdate
```

### Actual behavior

The Keycloak CRD reports its status using yaml booleans:

```yaml
status:
  conditions:
    - message: ''
      status: true
      type: Ready
    - message: ''
      status: false
      type: HasErrors
    - message: ''
      status: false
      type: RollingUpdate
```

### How to Reproduce?

Deploy Keycloak to kubernetes using the new operator and the Keycloak CRD.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by shawkins**
This aligns the condition with the typical kuberentes schema.  However it can be considered a breaking change as the crd type of an existing field is changing.  I'm not sure we really care, but:
- usage of the old model classes will fail to deserialized status=Unknown.  We aren't explicitly using Unknown so that's a minor concern.  Jackson will happily deserialize 'True' / 'False' as a boolean, so the base cases will continue to work even with Boolean in the model. 
- anyone who built a language binding off of the old crd may need to update that to the new crd because their parsing may not be as lax as Jackson.
- any reliance on the lower case true / false values in the status won't work, like the change needed to the check-examples-installed script.

Closes #13074

### CODE REVIEW

**Comment by nick4fake** [comment]
There might be some temporary flag which reverts behaviour to bools if anyone somehow depends on it.

**Comment by shawkins** [comment]
> There might be some temporary flag which reverts behaviour to bools if anyone somehow depends on it.

t'm not sure what that would look like.  The only other approach is to treat this as a truly breaking crd change that requires a new vision of the crd and provide a conversion webhook for the old version.  

Let me check into wha needs done with #21063 as well.

**Comment by shawkins** [comment]
@mabartos @Pepo48 any thoughts on leaving this as a small breaking change?

@abstractj added a note about this in the docs

File: docs/documentation/release_notes/topics/22_0_0.adoc
Comment (Pepo48) [inline docs/documentation/release_notes/topics/22_0_0.adoc:91]:
```suggestion
The are additional fields available in the keycloak.status to facilitate keycloak being a scalable resource. There are also additional fields that make the status easier to interpret such as observedGeneration and condition observedGeneration and lastTransitionTime fields. However the condition status field was also changed from a boolean to a string for conformance with standard Kubernetes conditions. Please make sure any of your usage of this field is updated to expect the values "True", "False", or "Unknown", rather than true or false.
```

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/StatusCondition.java
Comment (Pepo48) [inline operator/src/main/java/org/keycloak/operator/crds/v2alpha1/StatusCondition.java:27]: don't you want to add yourself as the author, since it's a newly added file?

**Review by Pepo48 — APPROVED** [review]
@shawkins I like the change. I would maybe wait for a blessing from Vašek regarding the breaking change (although it's minor), as he returns back on Monday. But if Bruno and Martin will be ok with merging it immediately, I won't be against 😊.

Also, I added two comments - very small cosmetic nitpicks that might or might not be addressed (I leave it up to you). Therefore, I'm approving it.

Thanks, good job!

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/StatusCondition.java
Comment (shawkins) [reply operator/src/main/java/org/keycloak/operator/crds/v2alpha1/StatusCondition.java:27]: I don't think I've ever used author tags before.  If that's a convention for the project, then I'll start adding them.

Comment (abstractj) [reply operator/src/main/java/org/keycloak/operator/crds/v2alpha1/StatusCondition.java:27]: That's not strictly necessary, as far as I'm aware, we don't have any conventions in this regard.

**Review by abstractj — APPROVED** [review]

**Comment by shawkins** [comment]
@vmuzikar raised that olm may be enforcing a validation against any possibly breaking crd changes to the current version.  This is coming from https://github.com/operator-framework/operator-lifecycle-manager/blob/08a95add7638746f205784b6f82c87983319151e/pkg/controller/operators/catalog/step.go#L137 

There seems to be no exceptions for status, so I need to try this against this change specifically and see if it runs afoul of the validation.  If it does, then this will need to be refined.  Either as a completely new crd version, or to a widening conversion inclusive of booleans, that we'll then narrow in a later release.

**Comment by shawkins** [comment]
> There seems to be no exceptions for status, so I need to try this against this change specifically and see if it runs afoul of the validation.

Confirmed that it does.  Also confirmed the workaround of widening.  So to do this without a new crd version would be to instead type status as AnyType or JsonNode for 22, then type it as string for a later version.

@vmuzikar do you have a preference on the solution?

**Comment by vmuzikar** [comment]
@shawkins Thanks for confirming this. That's what I was a bit afraid of. 😕

Introducing a new CRD just days before a release is from my perspective not an option, since we don't have this figured out, it'll be our first time.

Considering all options, I'm not sure I'm too much in favour of introducing another workaround/tech debt. Users would need to adapt to the breaking change of AnyType and soon after another breaking change of string. TBH, I'm more inclined to reverting this for now and introducing a proper new CRD version with string as status in perhaps 23. We also need to figure out our deprecation process first.

WDYT @shawkins @abstractj @Pepo48?

**Comment by vmuzikar** [comment]
@shawkins I was digging a bit more around the workaround. I agree from the user perspective, it truly seemingly acts as a string which is the desired behaviour. So not another breaking change, you are right.

I'm not very fond of the AnyType workaround but if we want align with K8s best practices (use string instead of boolean for status) and do not want to introduce a new CRD version at the same time, it is what it is.

**Pull request merged**
